calico_upgrade/pkg/clients/clients.go

```diff
@@ -0,0 +1,159 @@
+	cv3, err := clientv3.New(*v3ApiConfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error with apiconfigv3: %v", err)
+	}
+
+	// This must be an etcd backend.  Grab the Calico v1 API config (which must be specified).
+	// We'll need to convert to the v3 API config format to create the etcdv2 backend client.
+	v1ApiConfig, err := loadClientConfigV1(v1Config)
+	if v1ApiConfig.Spec.DatastoreType != apiv1.EtcdV2 {
+		return nil, nil, fmt.Errorf("expecting apiconfigv1 datastore to be 'etcdv2', got '%s'", v1ApiConfig.Spec.DatastoreType)
+	}
+	// Create the backend etcdv2 client (v1 API).  We wrap this in the compat module to handle
+	// multi-key backed resources.
+	ev1, err := etcdv2.NewEtcdClient(&v1ApiConfig.Spec.EtcdConfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error with apiconfigv1: %v", err)
+	}
+	bv1 := compat.NewAdaptor(ev1)
+
+	return cv3, bv1, nil
+}
+
+// loadClientConfigV1 loads the ClientConfig from the specified file (if specified)
+// or from environment variables (if the file is not specified).
+func loadClientConfigV1(filename string) (*apiv1.CalicoAPIConfig, error) {
+
+	// Override / merge with values loaded from the specified file.
+	if filename != "" {
+		b, err := ioutil.ReadFile(filename)
+		if err != nil {
+			return nil, err
+		}
+
+		c, err := loadClientConfigFromBytesV1(b)
+		if err != nil {
+			return nil, fmt.Errorf("syntax error in %s: %s", filename, err)
+		}
+		return c, nil
+	} else {
+		return loadClientConfigFromEnvironmentV1()
+	}
+}
+
+// loadClientConfigFromBytesV1 loads the ClientConfig from the supplied bytes containing
+// YAML or JSON format data.
+func loadClientConfigFromBytesV1(b []byte) (*apiv1.CalicoAPIConfig, error) {
+	var c apiv1.CalicoAPIConfig
+
+	// Default the backend type to be etcd v2.  This will be overridden if
+	// explicitly specified in the file.
+	log.Info("Loading config from JSON or YAML data")
+	c = apiv1.CalicoAPIConfig{
+		Spec: apiv1.CalicoAPIConfigSpec{
+			DatastoreType: apiv1.EtcdV2,
+		},
+	}
+
+	if err := yaml.UnmarshalStrict(b, &c); err != nil {
+		return nil, err
+	}
+
+	// Validate the version and kind.
+	if c.APIVersion != unversioned.VersionCurrent {
+		return nil, errors.New("invalid config file: unknown APIVersion '" + c.APIVersion + "'")
+	}
+	if c.Kind != "calicoApiConfig" {
+		return nil, errors.New("invalid config file: expected kind 'calicoApiConfig', got '" + c.Kind + "'")
+	}
+
+	log.Info("Datastore type: ", c.Spec.DatastoreType)
+	return &c, nil
+}
+
+// loadClientConfigFromEnvironmentV1 loads the ClientConfig from the specified file (if specified)
+// or from environment variables (if the file is not specified).
+func loadClientConfigFromEnvironmentV1() (*apiv1.CalicoAPIConfig, error) {
+	c := apiv1.NewCalicoAPIConfig()
+
+	// Load client config from environment variables.
+	log.Info("Loading config from environment")
+	if err := envconfig.Process("APIV1", &c.Spec); err != nil {
```

Comment:
v1 client use to default ETCD endpoint to localhost, will that still be the case? if so, should we fail unless it's provided for v1?

Reply:
I've removed the default.  Everything needs to be explicit for the upgrade script.